source/distributed/full_distributed_tria.cc

```diff
@@ -0,0 +1,499 @@
+                  for (; fine_cell_info != cell_infos[ref_counter].end();
+                       ++fine_cell_info)
+                    {
+                      // find the parent of that cell
+                      while (!internal::is_parent<dim>(
+                        coarse_cell->id().template to_binary<dim>(),
+                        fine_cell_info->id))
+                        coarse_cell++;
+
+                      // set parent for refinement
+                      coarse_cell->set_refine_flag();
+                    }
+
+                  // execute refinement
+                  dealii::Triangulation<dim, spacedim>::
+                    execute_coarsening_and_refinement();
+                }
+
+              // b) set manifold ids (because new vertices have to be positioned
+              //    correctly)
+              {
+                auto cell      = this->begin(ref_counter);
+                auto cell_info = cell_infos[ref_counter].begin();
+                for (; cell_info != cell_infos[ref_counter].end(); ++cell_info)
+                  {
+                    while (cell_info->id !=
+                           cell->id().template to_binary<dim>())
+                      cell++;
+                    if (spacedim == 3)
+                      for (unsigned int quad = 0;
+                           quad < GeometryInfo<spacedim>::quads_per_cell;
+                           quad++)
+                        cell->quad(quad)->set_manifold_id(
+                          cell_info->manifold_quad_ids[quad]);
+
+                    if (spacedim >= 2)
+                      for (unsigned int line = 0;
+                           line < GeometryInfo<spacedim>::lines_per_cell;
+                           line++)
+                        cell->line(line)->set_manifold_id(
+                          cell_info->manifold_line_ids[line]);
+
+                    cell->set_manifold_id(cell_info->manifold_id);
+                  }
+              }
+            }
+
+
+
+          // 4a) set all cells artificial
+          for (auto cell = this->begin(); cell != this->end(); cell++)
+            {
+              if (cell->active())
+                cell->set_subdomain_id(
+                  dealii::numbers::artificial_subdomain_id);
+
+              cell->set_level_subdomain_id(
+                dealii::numbers::artificial_subdomain_id);
+            }
+
+
+
+          // 4b) set actual (level_)subdomain_ids as well as boundary ids
+          for (unsigned int ref_counter = 0; ref_counter < cell_infos.size();
+               ref_counter++)
+            {
+              auto cell      = this->begin(ref_counter);
+              auto cell_info = cell_infos[ref_counter].begin();
+              for (; cell_info != cell_infos[ref_counter].end(); ++cell_info)
+                {
+                  // find cell that has the correct
+                  while (cell_info->id != cell->id().template to_binary<dim>())
```

Comment:
What happens if you hand in incorrect data and we don't find the cell. Is this an endless loop? Maybe you can check if we arrive at end()?

Reply:
There is not endless loop. Instead, the cell iterator (reaches its end and) tries to access an invalid cell, which is the time an assert is encountered and an assert is thrown.